File: kapesit/quantum/python/communication/quantum_communication.py
import numpy as np
from typing import Tuple, List, Dict
import threading
from dataclasses import dataclass
from concurrent.futures import ThreadPoolExecutor

@dataclass
class Complex:
    real: float
    imag: float

    def __mul__(self, other: 'Complex') -> 'Complex':
        return Complex(
            self.real * other.real - self.imag * other.imag,
            self.real * other.imag + self.imag * other.real
        )

    def __add__(self, other: 'Complex') -> 'Complex':
        return Complex(self.real + other.real, self.imag + other.imag)

    def magnitude(self) -> float:
        return np.sqrt(self.real * self.real + self.imag * self.imag)

class QuantumCommunication:
    def __init__(self, num_qubits: int):
        self.num_qubits = num_qubits
        self._lock = threading.RLock()
        self._gates = self._initialize_gates()
        self._executor = ThreadPoolExecutor(max_workers=4)

    def _initialize_gates(self) -> Dict[str, np.ndarray]:
        """Initialize quantum gates."""
        sqrt2 = 1.0 / np.sqrt(2)
        
        # Hadamard gate
        h_gate = np.array([
            [Complex(sqrt2, 0), Complex(sqrt2, 0)],
            [Complex(sqrt2, 0), Complex(-sqrt2, 0)]
        ])

        # CNOT gate
        cnot_gate = np.array([
            [Complex(1, 0), Complex(0, 0), Complex(0, 0), Complex(0, 0)],
            [Complex(0, 0), Complex(1, 0), Complex(0, 0), Complex(0, 0)],
            [Complex(0, 0), Complex(0, 0), Complex(0, 0), Complex(1, 0)],
            [Complex(0, 0), Complex(0, 0), Complex(1, 0), Complex(0, 0)]
        ])

        return {
            'H': h_gate,
            'CNOT': cnot_gate
        }

    def generate_entangled_pair(self) -> Tuple[np.ndarray, np.ndarray]:
        """Generate a Bell state between two qubits."""
        with self._lock:
            state1 = np.zeros(2, dtype=object)
            state2 = np.zeros(2, dtype=object)

            # Initialize to |0⟩
            state1[0] = Complex(1, 0)
            state2[0] = Complex(1, 0)

            # Apply Hadamard to first qubit
            self._apply_gate('H', 0, state1)

            # Apply CNOT
            self._apply_cnot(state1, state2)

            return state1, state2

    def quantum_teleport(self, state: np.ndarray, entangled_state: np.ndarray) -> np.ndarray:
        """Implement quantum teleportation protocol."""
        if len(state) != 2 or len(entangled_state) != 2:
            raise ValueError("States must be single qubits")

        with self._lock:
            # Apply CNOT between state and first qubit of entangled pair
            self._apply_cnot(state, entangled_state)

            # Apply Hadamard to state
            self._apply_gate('H', 0, state)

            # Measure both qubits
            measurement1 = self._measure(state)
            measurement2 = self._measure(entangled_state)

            return np.array([measurement1, measurement2], dtype=np.uint8)

    def quantum_dense_coding(self, message: int, entangled_state: np.ndarray) -> np.ndarray:
        """Implement dense coding protocol."""
        if len(entangled_state) != 2:
            raise ValueError("State must be a single qubit")

        with self._lock:
            # Apply operations based on message
            if message == 1:
                self._apply_gate('X', 0, entangled_state)
            elif message == 2:
                self._apply_gate('Z', 0, entangled_state)
            elif message == 3:
                self._apply_gate('X', 0, entangled_state)
                self._apply_gate('Z', 0, entangled_state)

            return entangled_state

    def quantum_dense_decoding(self, state1: np.ndarray, state2: np.ndarray) -> int:
        """Decode a dense coded message."""
        if len(state1) != 2 or len(state2) != 2:
            raise ValueError("States must be single qubits")

        with self._lock:
            # Apply CNOT
            self._apply_cnot(state1, state2)

            # Apply Hadamard to first qubit
            self._apply_gate('H', 0, state1)

            # Measure both qubits
            measurement1 = self._measure(state1)
            measurement2 = self._measure(state2)

            # Combine measurements to get original message
            return (measurement1 << 1) | measurement2

    def _apply_gate(self, gate_name: str, qubit: int, state: np.ndarray) -> None:
        """Apply a quantum gate to a qubit."""
        gate = self._gates.get(gate_name)
        if gate is None:
            raise ValueError(f"Unknown gate: {gate_name}")

        new_state = np.zeros_like(state)
        for i in range(len(state)):
            bit = (i >> qubit) & 1
            idx = i ^ (1 << qubit)
            if i < idx:
                new_state[i] = (gate[0][0] * state[i] + 
                              gate[0][1] * state[idx])
                new_state[idx] = (gate[1][0] * state[i] + 
                                gate[1][1] * state[idx])

        state[:] = new_state

    def _apply_cnot(self, control: np.ndarray, target: np.ndarray) -> None:
        """Apply CNOT gate between control and target qubits."""
        new_control = np.zeros_like(control)
        new_target = np.zeros_like(target)

        for i in range(len(control)):
            for j in range(len(target)):
                control_bit = (i >> 0) & 1
                target_bit = (j >> 0) & 1
                new_target_bit = target_bit ^ control_bit

                if control_bit == 1:
                    new_control[i] = control[i]
                    new_target[new_target_bit] = target[j]
                else:
                    new_control[i] = control[i]
                    new_target[j] = target[j]

        control[:] = new_control
        target[:] = new_target

    def _measure(self, state: np.ndarray) -> int:
        """Measure a quantum state."""
        probabilities = np.array([s.magnitude() ** 2 for s in state])
        r = np.random.random()
        result = 0 if r < probabilities[0] else 1

        # Collapse state
        new_state = np.zeros_like(state)
        new_state[result] = Complex(1, 0)
        state[:] = new_state

        return result

    def __del__(self):
        """Cleanup resources."""
        self._executor.shutdown(wait=True) 

File: kapesit/quantum/csharp/communication/QuantumCommunication.cs
using System;
using System.Numerics;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Kapesit.Quantum.Communication
{
    public class QuantumCommunication
    {
        private readonly Random random;
        private readonly Dictionary<string, Complex[,]> gates;
        private readonly int numQubits;

        public QuantumCommunication(int numQubits)
        {
            this.numQubits = numQubits;
            this.random = new Random();
            this.gates = InitializeGates();
        }

        private Dictionary<string, Complex[,]> InitializeGates()
        {
            var gates = new Dictionary<string, Complex[,]>();
            
            // Hadamard gate
            var sqrt2 = 1.0 / Math.Sqrt(2);
            gates["H"] = new Complex[,] {
                { new Complex(sqrt2, 0), new Complex(sqrt2, 0) },
                { new Complex(sqrt2, 0), new Complex(-sqrt2, 0) }
            };

            // CNOT gate
            gates["CNOT"] = new Complex[,] {
                { Complex.One, Complex.Zero, Complex.Zero, Complex.Zero },
                { Complex.Zero, Complex.One, Complex.Zero, Complex.Zero },
                { Complex.Zero, Complex.Zero, Complex.Zero, Complex.One },
                { Complex.Zero, Complex.Zero, Complex.One, Complex.Zero }
            };

            return gates;
        }

        public (Complex[] state1, Complex[] state2) GenerateEntangledPair()
        {
            var state1 = new Complex[2];
            var state2 = new Complex[2];

            // Initialize to |0⟩
            state1[0] = Complex.One;
            state2[0] = Complex.One;

            // Apply Hadamard to first qubit
            ApplyGate("H", 0, state1);

            // Apply CNOT
            ApplyCNOT(state1, state2);

            return (state1, state2);
        }

        public byte[] QuantumTeleport(Complex[] state, Complex[] entangledState)
        {
            if (state.Length != 2 || entangledState.Length != 2)
                throw new ArgumentException("States must be single qubits");

            // Apply CNOT between state and first qubit of entangled pair
            ApplyCNOT(state, entangledState);

            // Apply Hadamard to state
            ApplyGate("H", 0, state);

            // Measure both qubits
            int measurement1 = Measure(state);
            int measurement2 = Measure(entangledState);

            // Apply corrections based on measurements
            var result = new byte[2];
            result[0] = (byte)measurement1;
            result[1] = (byte)measurement2;

            return result;
        }

        public Complex[] QuantumDenseCoding(byte message, Complex[] entangledState)
        {
            if (entangledState.Length != 2)
                throw new ArgumentException("State must be a single qubit");

            // Apply operations based on message
            switch (message)
            {
                case 0:
                    // Do nothing
                    break;
                case 1:
                    ApplyGate("X", 0, entangledState);
                    break;
                case 2:
                    ApplyGate("Z", 0, entangledState);
                    break;
                case 3:
                    ApplyGate("X", 0, entangledState);
                    ApplyGate("Z", 0, entangledState);
                    break;
            }

            return entangledState;
        }

        public byte QuantumDenseDecoding(Complex[] state1, Complex[] state2)
        {
            if (state1.Length != 2 || state2.Length != 2)
                throw new ArgumentException("States must be single qubits");

            // Apply CNOT
            ApplyCNOT(state1, state2);

            // Apply Hadamard to first qubit
            ApplyGate("H", 0, state1);

            // Measure both qubits
            int measurement1 = Measure(state1);
            int measurement2 = Measure(state2);

            // Combine measurements to get original message
            return (byte)((measurement1 << 1) | measurement2);
        }

        private void ApplyGate(string gateName, int qubit, Complex[] state)
        {
            if (!gates.TryGetValue(gateName, out Complex[,] gate))
                throw new ArgumentException($"Unknown gate: {gateName}");

            var newState = new Complex[state.Length];
            for (int i = 0; i < state.Length; i++)
            {
                int bit = (i >> qubit) & 1;
                int idx = i ^ (1 << qubit);
                if (i < idx)
                {
                    newState[i] = gate[0, 0] * state[i] + gate[0, 1] * state[idx];
                    newState[idx] = gate[1, 0] * state[i] + gate[1, 1] * state[idx];
                }
            }
            Array.Copy(newState, state, state.Length);
        }

        private void ApplyCNOT(Complex[] control, Complex[] target)
        {
            var newControl = new Complex[control.Length];
            var newTarget = new Complex[target.Length];

            for (int i = 0; i < control.Length; i++)
            {
                for (int j = 0; j < target.Length; j++)
                {
                    int controlBit = (i >> 0) & 1;
                    int targetBit = (j >> 0) & 1;
                    int newTargetBit = targetBit ^ controlBit;

                    if (controlBit == 1)
                    {
                        newControl[i] = control[i];
                        newTarget[newTargetBit] = target[j];
                    }
                    else
                    {
                        newControl[i] = control[i];
                        newTarget[j] = target[j];
                    }
                }
            }

            Array.Copy(newControl, control, control.Length);
            Array.Copy(newTarget, target, target.Length);
        }

        private int Measure(Complex[] state)
        {
            double[] probabilities = new double[2];
            for (int i = 0; i < state.Length; i++)
            {
                probabilities[i] = state[i].Magnitude * state[i].Magnitude;
            }

            double r = random.NextDouble();
            int result = r < probabilities[0] ? 0 : 1;

            // Collapse state
            var newState = new Complex[state.Length];
            newState[result] = Complex.One;
            Array.Copy(newState, state, state.Length);

            return result;
        }
    }
} 